File: examples/zig/terminal-varlen-read/TerminalVarlenRead.runar.zig
const runar = @import("runar");

pub const TerminalVarlenRead = struct {
    pub const Contract = runar.StatefulSmartContract;

    message: runar.ByteString = "",

    pub fn init(message: runar.ByteString) TerminalVarlenRead {
        return .{ .message = message };
    }

    pub fn post(self: *TerminalVarlenRead, newMessage: runar.ByteString) void {
        self.message = newMessage;
    }

    pub fn reveal(self: *const TerminalVarlenRead, minLen: i64) void {
        runar.assert(runar.len(self.message) > minLen);
    }
};

File: examples/java/src/main/java/runar/examples/terminal-varlen-read/TerminalVarlenRead.runar.java
package runar.examples.terminalvarlenread;

import runar.lang.StatefulSmartContract;
import runar.lang.annotations.Public;
import runar.lang.types.Bigint;
import runar.lang.types.ByteString;

import static runar.lang.Builtins.assertThat;
import static runar.lang.Builtins.len;

/**
 * TerminalVarlenRead -- regression fixture for GitHub issue #100: a terminal
 * method that reads the mutable variable-length (ByteString) state field.
 */
class TerminalVarlenRead extends StatefulSmartContract {

    ByteString message;

    TerminalVarlenRead(ByteString message) {
        super(message);
        this.message = message;
    }

    @Public
    void post(ByteString newMessage) {
        this.message = newMessage;
    }

    @Public
    void reveal(Bigint minLen) {
        assertThat(len(this.message).gt(minLen));
    }
}
